src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs

```diff
@@ -1276,5 +1304,57 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
             result = null;
             return false;
         }
+
+        private static class TemplateHelper
+        {
+            private const string MapTemplateStartDelimiter = "#{{";
+            private const string MapTemplateEndDelimiter = "}}#";
+            private const string MapTemplate = MapTemplateStartDelimiter + "{{{0}}}" + MapTemplateEndDelimiter;
+
+            private static string GetNotReplacedTemplate(string value)
+            {
+                var startIndexOf = value.IndexOf(MapTemplateStartDelimiter, StringComparison.Ordinal);
+                var endIndexOf = value.IndexOf(MapTemplateEndDelimiter, StringComparison.Ordinal);
+                if (startIndexOf != -1 && endIndexOf != -1 && endIndexOf > startIndexOf)
+                {
+                    endIndexOf = endIndexOf + MapTemplateEndDelimiter.Length;
+                    if (endIndexOf <= value.Length) return value.Substring(startIndexOf, endIndexOf - startIndexOf);
+                }
+
+                return null;
+            }
+
+            public static BsonValue ReplaceTemplates(BsonValue bsonValue, Dictionary<string,string> mapHistory)
```

Comment:
During `Translation` processing, in low-level expressions, we don't have information about mapping ($map, $filter) nodes, which will be generated in higher levels. So, in translation steps, we don't make any mappings. We just mark these nodes as `template`s which contain `key` values. In this method, these templates will be replaced on final query values, based on prepared _mapHistory.